packages/starknet-snap/src/ui/components/ExecuteTxnUI.tsx

```diff
@@ -0,0 +1,207 @@
+import type { SnapComponent } from '@metamask/snaps-sdk/jsx';
+import {
+  Box,
+  Container,
+  Row,
+  Section,
+  Text,
+  Icon,
+  Divider,
+} from '@metamask/snaps-sdk/jsx';
+import { formatUnits } from 'ethers/lib/utils';
+
+import type { FeeToken } from '../../types/snapApi';
+import type { FormattedCallData } from '../../types/snapState';
+import { DEFAULT_DECIMAL_PLACES } from '../../utils/constants';
+import { AddressUI, JsonDataUI } from '../fragments';
+import { FeeTokenSelector } from '../fragments/FeeTokenSelector';
+
+export type ExecuteTxnUIProps = {
+  type: string;
+  signer: string;
+  chainId: string;
+  maxFee: string;
+  calls: FormattedCallData[];
+  selectedFeeToken: string;
+  includeDeploy: boolean;
+};
+
+type TokenTotals = Record<
+  string,
+  {
+    amount: bigint; // Use BigInt for precise calculations
+    decimals: number;
+  }
+>;
+
+/**
+ * Calculates the totals for all tokens involved in calls and fees.
+ *
+ * @param calls - The transaction calls.
+ * @param maxFee - The maximum fee as a string BigInt.
+ * @param selectedFeeToken - The token symbol used for fees.
+ * @returns The calculated totals for each token.
+ */
+export const calculateTotals = (
+  calls: {
+    isTransfer: boolean;
+    transferAmount?: string; // Amount as a string BigInt
+    transferTokenSymbol?: string;
+    transferTokenDecimals?: number;
+  }[],
+  maxFee: string,
+  selectedFeeToken: string,
+): TokenTotals => {
+  const tokenTotals: TokenTotals = {};
+
+  // Sum up transfer amounts for each token
+  calls.forEach((call) => {
+    if (
+      call.transferAmount &&
+      call.transferTokenSymbol &&
+      call.transferTokenDecimals &&
+      call.isTransfer
+    ) {
+      const amount = BigInt(call.transferAmount); // Convert to BigInt
+      if (!tokenTotals[call.transferTokenSymbol]) {
+        tokenTotals[call.transferTokenSymbol] = {
+          amount: BigInt(0),
+          decimals: call.transferTokenDecimals,
+        };
+      }
+      tokenTotals[call.transferTokenSymbol].amount += amount;
+    }
+  });
+
+  // Add the fee to the corresponding token
+  const feeTokenAmount = BigInt(maxFee);
+  if (tokenTotals[selectedFeeToken]) {
+    tokenTotals[selectedFeeToken].amount += feeTokenAmount;
+  } else {
+    // We derive decimals based on the fee token. Currently, both supported fee tokens, ETH and STRK, use the standard 18 decimals.
+    // Therefore, we use DEFAULT_DECIMAL_PLACES set to 18 here. If additional fee tokens with different decimals are introduced,
+    // this logic should be updated to handle token-specific decimals dynamically.
+    tokenTotals[selectedFeeToken] = {
+      amount: feeTokenAmount,
+      decimals: DEFAULT_DECIMAL_PLACES,
+    };
+  }
+
```

Comment:
nit:

i think we can simplify the code in that way

```suggestion
  const tokenTotals: TokenTotals = calls.reduce((acc, call) => {
    if (
      call.transferAmount &&
      call.transferTokenSymbol &&
      call.transferTokenDecimals &&
      call.isTransfer
    ) {
      const amount = BigInt(call.transferAmount); // Convert to BigInt
      if (!acc[call.transferTokenSymbol]) {
        acc[call.transferTokenSymbol] = {
          amount: BigInt(0),
          decimals: call.transferTokenDecimals,
        };
      }
      acc[call.transferTokenSymbol].amount += amount;
    }
    return acc;
  }, {
    // We derive decimals based on the fee token. Currently, both supported fee tokens, ETH and STRK, use the standard 18 decimals.
    // Therefore, we use DEFAULT_DECIMAL_PLACES set to 18 here. If additional fee tokens with different decimals are introduced,
    // this logic should be updated to handle token-specific decimals dynamically.
    [selectedFeeToken]: {
      amount: BigInt(maxFee),
      decimals: DEFAULT_DECIMAL_PLACES,
    }
  })
```